runtime/pkg/container/container.go

```diff
@@ -0,0 +1,118 @@
+package container
+
+import (
+	"container/list"
+	"fmt"
+)
+
+// Container is a collection of items
+// More functions like remove can be added if future use cases arise
+type Container[T any] interface {
+	// Add consumes an item
+	Add(item T) bool
+
+	// IsFull return true if container is full
+	IsFull() bool
+
+	// returns items as slice
+	Items() []T
+}
+
+// boundedContainer is a container with limited capacity
+// Stops consuming more elements once capacity is reached
+// This is thread unsafe
+type boundedContainer[T any] struct {
+	items []T
+	index int
+}
+
+func NewBoundedContainer[T any](capacity int) (Container[T], error) {
+	if capacity <= 0 {
+		return nil, fmt.Errorf("len should be greater than 0")
+	}
+	b := &boundedContainer[T]{items: make([]T, capacity), index: 0}
+	return b, nil
+}
+
+func (b *boundedContainer[T]) Add(item T) bool {
+	if b.IsFull() {
+		return false
+	}
+
+	b.items[b.index] = item
+	b.index++
+	return true
+}
+
+func (b *boundedContainer[T]) IsFull() bool {
+	return b.index == len(b.items)
+}
+
+func (b *boundedContainer[T]) Items() []T {
+	return b.items[:b.index]
+}
+
+// tailContainer is a container with limited capacity
+// Keeps last 'N' inserted elements upto capacity
+// This is thread unsafe
+type tailContainer[T any] struct {
+	items     *list.List
+	capacity  int
+	cleanupfn func(T)
+}
+
+func NewTailContainer[T any](capacity int, cleanupfn func(T)) (Container[T], error) {
```

Comment:
1. Consider `FIFO` instead of `tail` as a generic expression
2. Only call `cleanupFn` if it is not nil, and pass `nil` in places where it doesn't do anything (instead of empty function)